File: ImprovedPooledLogit/pooled_logit.R
# Specifically readr=2.1.4, dplyr=1.1.1, tidyr=1.3.0,
if (!require("tidyverse")) {
  install.packages('tidyverse')
}
library(tidyverse)

if (!require("microbenchmark")) {
  install.packages('microbenchmark')
}
library(microbenchmark)


# Read in dataset - then subset
df = read_csv("lau.csv") |>
  mutate(
    t_days = ceiling(365.25 * t),
    event_date = t_days,
  ) |>
  uncount(t_days, .id="days") |>
  mutate(
    event_indicator = ((event_date - days) == 0) & (eventtype == 2)
    )

# Tidyverse-friendly way to find all unique event times, as a vector
unique_event_times = df |> filter(event_indicator) |> pull(days) |> unique()


# Create the 1y dataset - filter by days, and optionally filter by being at a unique event time
df_1y = df |> filter(days <= 365.25)
df_1y_improved = df |>
  filter(days <= 365.25) |>
  filter(days %in% unique_event_times)

# Create the 5y dataset - same way as 1y dataset
df_5y = df |> filter(days <= 5 * 365.25)
df_5y_improved = df |>
  filter(days <= 5 * 365.25) |>
  filter(days %in% unique_event_times)


# Improved implementation
model_1y_improved = glm(
  event_indicator ~ factor(days) + ageatfda + BASEIDU + black + cd4nadir,
  data = df_1y_improved,
  family=binomial(link="logit")
)

# Benchmark of 1-year improved implementation with 100 trials
mb_1y_improved = microbenchmark(
  glm(
    event_indicator ~ factor(days) + ageatfda + BASEIDU + black + cd4nadir,
    data = df_1y_improved,
    family=binomial(link="logit")
  ),
  times=100
)

# Benchamrk of 1-year standard implementation with 100 trials
mb_1y = microbenchmark(
  glm(
    event_indicator ~ factor(days) + ageatfda + BASEIDU + black + cd4nadir,
    data = df_1y,
    family=binomial(link="logit")
  ),
  times=100
)

# Benchmark of 5-year improved implementation with 100 trials
mb_5y_improved = microbenchmark(
  glm(
    event_indicator ~ factor(days) + ageatfda + BASEIDU + black + cd4nadir,
    data = df_5y_improved,
    family=binomial(link="logit")
  ),
  times=100
)

# Benchmark of 5-year standard implementation with 100 trials. Takes a LONG time.
mb_5y = microbenchmark(
  glm(
    event_indicator ~ factor(days) + ageatfda + BASEIDU + black + cd4nadir,
    data = df_5y,
    family=binomial(link="logit")
  ),
  times=100
)

File: ImprovedPooledLogit/pooled_logit.py
import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf
from time import time

max_follow_up_years = 5

##########################################
# Setup data

d = pd.read_csv("lau.csv")

# Recoding events and admin censoring
d['event'] = np.where(d['eventtype'] == 2, 1, 0)                              # Competing event is event of interest
d['event'] = np.where(d['t'] > max_follow_up_years, 0, d['event'])            # Administrative censoring at ...
d['t'] = np.where(d['t'] > max_follow_up_years, max_follow_up_years, d['t'])  # ... 5 years of follow-up

# Transforming time into days
d['days'] = np.ceil(d['t'] * 365.25)
d['days'] = d['days'].astype(int)
d.info()

##########################################
# Standard Implementation

#############
# Long data set
max_t = np.max(d['days'])
dl = pd.DataFrame(np.repeat(d.values, max_t, axis=0), columns=d.columns)
dl['t_in'] = dl.groupby("id")['days'].cumcount()
dl['t_out'] = dl['t_in'] + 1

dl['delta'] = np.where(dl['t_out'] == dl['days'], dl['event'], 0)
dl['delta'] = np.where(dl['t_out'] > dl['days'], np.nan, dl['delta'])
dl = dl[['id', 't_out', 'delta', 'ageatfda', 'BASEIDU', 'black', 'cd4nadir']]
dl = dl.dropna(subset=['delta', ])
dl.info()

#############
# Specify model
model = "delta ~ BASEIDU + ageatfda + black + cd4nadir + C(t_out)"

#############
# Fit model
run_times = []
for i in range(100):
    start = time()
    fam = sm.families.Binomial()
    fm = smf.glm(model, data=dl, family=fam).fit()
    run_times.append(time() - start)

print("RUNTIMES:", np.min(run_times), np.mean(run_times), np.max(run_times))


##########################################
# Proposed Implementation

#############
# Long data set
max_t = np.max(d['days'])
dl = pd.DataFrame(np.repeat(d.values, max_t, axis=0), columns=d.columns)
dl['t_in'] = dl.groupby("id")['days'].cumcount()
dl['t_out'] = dl['t_in'] + 1
dl['delta'] = np.where(dl['t_out'] == dl['days'], dl['event'], 0)
dl['delta'] = np.where(dl['t_out'] > dl['days'], np.nan, dl['delta'])
dl = dl[['id', 't_out', 'delta', 'ageatfda', 'BASEIDU', 'black', 'cd4nadir']].copy()
dl = dl.dropna(subset=['delta', ])

#############
# Restrict

unique_event_times = np.unique(d.loc[d['event'] == 1, 'days'])
dl = dl.loc[dl['t_out'].isin(unique_event_times)].copy()
dl.info()

#############
# Specify model

model = "delta ~ BASEIDU + ageatfda + black + cd4nadir + C(t_out)"

#############
# Fit model
run_times = []
for i in range(100):
    start = time()
    fam = sm.families.Binomial()
    fm = smf.glm(model, data=dl, family=fam).fit()
    run_times.append(time() - start)

print("RUNTIMES:", np.min(run_times), np.mean(run_times), np.max(run_times))
